substrate/frame/revive/src/metering/math.rs

```diff
@@ -0,0 +1,526 @@
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+	BalanceOf, CallResources, Config, DispatchError, Error, EthTxInfo, FixedPointNumber, FixedU128,
+	FrameMeter, InfoT, ResourceMeter, RootStorageMeter, SaturatedConversion, Saturating, SignedGas,
+	State, StorageDeposit, TransactionLimits, TransactionMeter, Weight, WeightMeter, Zero,
+};
+use crate::{limits::SOLIDITY_CALL_STIPEND, metering::weight::Token, vm::evm::EVMGas};
+use core::marker::PhantomData;
+
+fn determine_call_stipend<T: Config>() -> Weight {
+	let gas = EVMGas(SOLIDITY_CALL_STIPEND.into());
+	<EVMGas as Token<T>>::weight(&gas)
+}
+
+pub mod substrate_execution {
+	use num_traits::One;
+
+	use super::*;
+
+	/// Create a transaction-level (root) meter for Substrate-style execution.
+	///
+	/// This constructs the root resource meter that enforces explicit weight and
+	/// storage-deposit limits for the whole transaction. The returned `TransactionMeter`:
+	/// - charges weight via `WeightMeter` with the provided `weight_limit`,
+	/// - accounts storage deposit via `RootStorageMeter` with the provided `deposit_limit`,
+	/// - records that the transaction's limit mode is `WeightAndDeposit`.
+	pub fn new_root<T: Config>(
+		weight_limit: Weight,
+		deposit_limit: BalanceOf<T>,
+	) -> Result<TransactionMeter<T>, DispatchError> {
+		Ok(TransactionMeter {
+			weight: WeightMeter::new(Some(weight_limit), None),
+			deposit: RootStorageMeter::new(Some(deposit_limit)),
+			// ignore max total gas for Substrate executions
+			max_total_gas: Default::default(),
+			total_consumed_weight_before: Default::default(),
+			total_consumed_deposit_before: Default::default(),
+			transaction_limits: TransactionLimits::WeightAndDeposit { weight_limit, deposit_limit },
+			_phantom: PhantomData,
+		})
+	}
+
+	/// Create a nested (frame) meter derived from a parent `ResourceMeter`.
+	///
+	/// This produces a frame-local meter that enforces the resource limits for
+	/// a nested call. It computes how much of the parent's remaining resources are available
+	/// to the nested frame by:
+	/// - collecting the parent's own consumed amounts (`self_consumed_*`),
+	/// - deriving the total consumed amounts up to this point,
+	/// - applying the requested `CallResources` (no limits, ethereum gas conversion, or explicit
+	///   weight+deposit) to derive per-frame limits.
+	///
+	/// Returns `Err(Error::OutOfGas)` when weight is exhausted, or
+	/// `Err(Error::StorageDepositLimitExhausted)` when deposit bookkeeping forbids further storage.
+	pub fn new_nested_meter<T: Config, S: State>(
+		meter: &ResourceMeter<T, S>,
```

Comment:
nit:  I would have expected that fn to be an impl on ResourceMeter and meter to be self, but if we keep it static, we can maybe rename it to `parent`